vo: hwdec: drmprime: add separate hwdecs for v4l2request

philipl opened the pull request:
With all the machinery in place, we can now add the v4l2request hwdecs with a different hw device type, and a different initialisation path. This applies to both the drmprime and drmprime_overlay hwdecs.

At the moment, the device initialisation is done in the bare minimum way, but it can be extended to take a device path (for example) if that makes sense as we better understand what meaningful configuration will be.

This change is based on the code proposed by Jonas Karlman <jonas@kwiboo.se> in https://github.com/mpv-player/mpv/pull/14511

philipl commented:
@Kwiboo I suggest we close #14511 and track the final support code in this PR which we'll merge once the ffmpeg changes are finalised.

diederikdehaas commented:
It's (very) likely just a dumb error on my part, but I'm getting an error when compiling it (on an arm64 device with Debian Trixie/Sid):

```sh
/usr/bin/ld: libmpv.so.2.3.0.p/video_out_gpu_hwdec.c.o:(.data.rel.ro+0x20): undefined reference to `ra_hwdec_v4l2request'
/usr/bin/ld: libmpv.so.2.3.0.p/video_out_gpu_hwdec.c.o:(.data.rel.ro+0x28): undefined reference to `ra_hwdec_v4l2request_overlay'
```

Parts of my build log (hopefully relevant):
```sh
make[1]: Entering directory '/home/diederik/dev/debian/salsa/multimedia-team/mpv'
dh_auto_configure -- \
	-Dlibmpv=true \
	-Dbuild-date=false \
	-Dcdda=enabled \
	-Ddvdnav=enabled \
	-Dsdl2=enabled \
	-Dv4l2request=enabled \
	-Davfoundation=disabled \
	-Degl-angle-win32=disabled \
	-Dgl-cocoa=disabled \
	-Dgl-win32=disabled \
	-Dios-gl=disabled \
	-Duwp=disabled \
	-Dvaapi-win32=disabled \
	-Dwasapi=disabled \
	-Dwin32-threads=disabled \
	-Ddvbin=enabled
	cd obj-aarch64-linux-gnu && DEB_PYTHON_INSTALL_LAYOUT=deb LC_ALL=C.UTF-8 meson setup .. --wrap-mode=nodownload --buildtype=plain --prefix=/usr --sysconfdir=/etc --localstatedir=/var --libdir=lib/aarch64-linux-gnu -Dpython.bytecompile=-1 -Dlibmpv=true -Dbuild-date=false -Dcdda=enabled -Ddvdnav=enabled -Dsdl2=enabled -Dv4l2request=enabled -Davfoundation=disabled -Degl-angle-win32=disabled -Dgl-cocoa=disabled -Dgl-win32=disabled -Dios-gl=disabled -Duwp=disabled -Dvaapi-win32=disabled -Dwasapi=disabled -Dwin32-threads=disabled -Ddvbin=enabled
The Meson build system
Version: 1.5.1
Source dir: /home/diederik/dev/debian/salsa/multimedia-team/mpv
Build dir: /home/diederik/dev/debian/salsa/multimedia-team/mpv/obj-aarch64-linux-gnu
Build type: native build
Project name: mpv
Project version: 0.38.0
C compiler for the host machine: cc (gcc 14.2.0 "cc (Debian 14.2.0-1) 14.2.0")
C linker for the host machine: cc ld.bfd 2.43
Host machine cpu family: aarch64
Host machine cpu: aarch64
...
Message: List of enabled features: alsa av-channel-layout avif-muxer caca cdda cplugins cuda-hwaccel cuda-interop dmabuf-interop-gl dmabuf-wayland drm dvbin dvdnav egl egl-drm egl-wayland egl-x11 ffmpeg ffnvcodec gbm gl glibc-thread-name glob glob-posix gpl iconv jack javascript jpeg jpegxl lavu-uuid lcms2 libarchive libass libavdevice libbluray libdl libplacebo linux-fstatfs lua52 memfd-create pipewire posix posix-shm ppoll pthread-condattr-setclock pulse rubberband rubberband-3 sdl2 sdl2-audio sdl2-gamepad sdl2-video sixel sndio sndio-1-9 uchardet v4l2request vaapi vaapi-drm vaapi-wayland vaapi-x11 vdpau vector vk-khr-display vt.h vulkan vulkan-interop wayland wayland-protocols-1-27 wayland-protocols-1-31 wayland-protocols-1-32 x11 xv zimg zimg-st428 zlib
Fetching value of define "MPV_CLIENT_API_VERSION" : (((2) << 16) | (3) | 0UL) 
Compiler for C supports link arguments -Wl,-Bsymbolic: YES 
Build targets in project: 46

mpv 0.38.0

    d3d11             : NO
    javascript        : YES
    libmpv            : YES
    lua               : YES
    opengl            : YES
    v4l2request       : YES
    vulkan            : YES
    wayland           : YES
    x11               : YES

  User defined options
    buildtype         : plain
    libdir            : lib/aarch64-linux-gnu
    localstatedir     : /var
    prefix            : /usr
    sysconfdir        : /etc
    wrap_mode         : nodownload
    python.bytecompile: -1
    avfoundation      : disabled
    build-date        : false
    cdda              : enabled
    dvbin             : enabled
    dvdnav            : enabled
    egl-angle-win32   : disabled
    gl-cocoa          : disabled
    gl-win32          : disabled
    ios-gl            : disabled
    libmpv            : true
    sdl2              : enabled
    uwp               : disabled
    v4l2request       : enabled
    vaapi-win32       : disabled
    wasapi            : disabled
    win32-threads     : disabled

Found ninja-1.12.1 at /usr/bin/ninja
```

What am I missing?  
Note: I was successful in building PR 14511

https://salsa.debian.org/diederik/mpv/-/commits/v4l2request-hwdevice is the source of my build

philipl commented:
@diederikdehaas yeah. I need to change how I did that conditional build logic.

Kwiboo commented on meson_options.txt:
```diff
@@ -101,6 +101,7 @@ option('d3d-hwaccel', type: 'feature', value: 'auto', description: 'D3D11VA hwac
 option('d3d9-hwaccel', type: 'feature', value: 'auto', description: 'DXVA2 hwaccel')
 option('gl-dxinterop-d3d9', type: 'feature', value: 'auto', description: 'OpenGL/DirectX DXVA2 hwaccel')
 option('ios-gl', type: 'feature', value: 'auto', description: 'iOS OpenGL ES interop support')
+option('v4l2request', type: 'feature', value: 'disabled', description: 'V4L2 Request API hwaccel')
```
```suggestion
option('v4l2request', type: 'feature', value: 'auto', description: 'V4L2 Request API hwaccel')
```

Kwiboo commented on video/out/hwdec/hwdec_drmprime.c:
```diff
@@ -114,45 +160,44 @@ static int init(struct ra_hwdec *hw)
-            MP_TARRAY_APPEND(p, p->formats, num_formats, pixfmt2imgfmt(fmt));
-        }
-    }
+#ifdef AV_HWDEVICE_TYPE_V4L2REQUEST
```
```suggestion
#if HAVE_V4L2REQUEST
```

Kwiboo commented on video/out/hwdec/hwdec_drmprime.c:
```diff
@@ -308,7 +353,24 @@ const struct ra_hwdec_driver ra_hwdec_drmprime = {
+    },
+};
+
+#ifdef AV_HWDEVICE_TYPE_V4L2REQUEST
```
```suggestion
#if HAVE_V4L2REQUEST
```

Kwiboo commented on video/out/hwdec/hwdec_drmprime_overlay.c:
```diff
@@ -324,12 +335,62 @@ static int init(struct ra_hwdec *hw)
     return -1;
 }
 
+#ifdef AV_HWDEVICE_TYPE_V4L2REQUEST
```
```suggestion
#if HAVE_V4L2REQUEST
```

Kwiboo commented on video/out/hwdec/hwdec_drmprime_overlay.c:
```diff
@@ -324,12 +335,62 @@ static int init(struct ra_hwdec *hw)
+    .uninit = uninit,
+};
+
+#ifdef AV_HWDEVICE_TYPE_V4L2REQUEST
```
```suggestion
#if HAVE_V4L2REQUEST
```

Kwiboo reviewed (changes requested):
`AV_HWDEVICE_TYPE_V4L2REQUEST` is a enum value so it cannot be used with `#ifdef`, an `#if` for `HAVE_V4L2REQUEST` should work.

Should also be possible to change the `v4l2request` feature to `auto` as it check if the `AV_HWDEVICE_TYPE_V4L2REQUEST` enum value exists.

diederikdehaas commented on video/out/hwdec/hwdec_drmprime_overlay.c:
```diff
@@ -324,12 +335,62 @@ static int init(struct ra_hwdec *hw)
+     *
+     * FIXME: Create hwdevice based on type in AVCodecHWConfig
+     */
+    int ret = av_hwdevice_ctx_create(&p->hwctx.av_device_ref,
```
video/out/hwdec/hwdec_drmprime_overlay.c:358:9: error: redefinition of ‘ret’ (of line 343)

diederikdehaas replied:
I applied @Kwiboo's suggestions (except from ``meson.options``) and fixed the above problem and then the package build successfully.  
But you need to do ``s/int ret/ret/``

Kwiboo committed:
meson: add detection logic for v4l2request support

We will probably adjust this to look for a specific libavutil version after
v4l2request support is merged upstream, but this check is fine for now.

philipl committed:
vo: hwdec: drmprime: add separate hwdecs for v4l2request

With all the machinery in place, we can now add the v4l2request hwdecs with a
different hw device type, and a different initialisation path. This applies to
both the drmprime and drmprime_overlay hwdecs.

At the moment, the device initialisation is done in the bare minimum way, but
it can be extended to take a device path (for example) if that makes sense as
we better understand what meaningful configuration will be.

Co-authored-by: Jonas Karlman <jonas@kwiboo.se>

kasper93 commented on meson.build:
```diff
@@ -1413,6 +1413,14 @@ if features['ios-gl']
     sources += files('video/out/hwdec/hwdec_ios_gl.m')
 endif
 
+v4l2request = get_option('v4l2request').require(
+    cc.has_header_symbol('libavutil/hwcontext.h', 'AV_HWDEVICE_TYPE_V4L2REQUEST')
+)
```
I would prefer to check lavu version once it is merged, instead of doing compile test everytime. It is also easier to track, which version is needed for this feature.

philipl replied:
Yes. Whether we merged this first or not, we would update this test to a version based one once it's in.

kasper93 commented on video/out/hwdec/hwdec_drmprime.c:
```diff
@@ -114,45 +160,44 @@ static int init(struct ra_hwdec *hw)
+    ret = av_hwdevice_ctx_create(&p->hwctx.av_device_ref,
+                                 AV_HWDEVICE_TYPE_V4L2REQUEST,
+                                 NULL, NULL, 0);
+    if (ret != 0) {
```
It is valid, but I think idiomatic check for ffmpeg errors is `ret < 0` generally.

kasper93 commented on video/out/hwdec/hwdec_drmprime.c:
```diff
@@ -114,45 +160,44 @@ static int init(struct ra_hwdec *hw)
+                                 NULL, NULL, 0);
+    if (ret != 0) {
+        MP_VERBOSE(hw, "Failed to create hwdevice_ctx: %s\n", av_err2str(ret));
+        return -1;
```
`return ret` would be ok too?

kasper93 commented on video/out/hwdec/hwdec_drmprime.c:
```diff
@@ -318,3 +380,4 @@ const struct ra_hwdec_driver ra_hwdec_drmprime = {
         .unmap = mapper_unmap,
     },
 };
+#endif
```
missing new line

kasper93 commented on video/out/hwdec/hwdec_drmprime_overlay.c:
```diff
@@ -324,12 +335,62 @@ static int init(struct ra_hwdec *hw)
     .overlay_frame = overlay_frame,
     .uninit = uninit,
 };
+#endif
```
missing new line

kasper93 reviewed (commented):
Look ok, not tested though. I would wait for ffmpeg side of changes to be merged and after that we can retest and merge this guy.

diederikdehaas commented:
> I would wait for ffmpeg side of changes to be merged and after that we can retest and merge this guy.

FWIW: I've verified that on my amd64 PC, which does not have the ffmpeg changes, the build 'still' succeeds, but just without v4l2request support:

```sh
Build targets in project: 46

mpv 0.38.0

    d3d11             : NO
    javascript        : YES
    libmpv            : YES
    lua               : YES
    opengl            : YES
    v4l2request       : NO
    vulkan            : YES
    wayland           : YES
    x11               : YES
```

And that is confirmed by a successful CI build here: https://salsa.debian.org/diederik/mpv/-/pipelines/718879  
(before autodetection I enabled it explicitly and then the CI build failed)

Whereas on my arm64 device which does have ffmpeg changes, the build succeeds as well, but also has this line:
```sh
    v4l2request       : YES
```

You can still choose to wait ofc.

Now on to runtime testing it...

diederikdehaas commented:
> Did you try with ``--vo=dmabuf-wayland``? Any difference?

With my/Debian based 0.38 + this patch set I noticed the following on my PineTab2 with Sway:

1. With videos where v4l2request is used, average CPU usage went from ~20% to ~10%
2. If I then request an OSD overlay (?) like with ``O`` or ``I`` and then press that again to make it disappear again, part of it remains on screen, which seems in the area of the black bars above (and probably below) the 'actual' video. When I go to normal screen size (and then full screen again), those left-overs disappear again
3. With some of my test videos, v4l2request doesn't work (same as with previous iterations of this patch set). When I use this (extra) parameter, I don't get a video at all, while without it I do see a video but this is then using CPU based rendering.

This could ofc already be fixed in ``master``, but I wanted to do my initial tests as close as possible to my previous ones.

Dudemanguy commented:
The second one has been reported in #13318. I was never personally able to reproduce but it's at least unrelated to the hwdec.

philipl commented:
> Look ok, not tested though. I would wait for ffmpeg side of changes to be merged and after that we can retest and merge this guy.

That's fine. I'll leave this year until it's in.

Headcrabed commented:
Besides, I think we have to wait for @Kwiboo 's v4l2request patch series land into mainline ffmpeg before we merge this?

diederikdehaas commented:
Yep.   
But for ppl like me who (self) build both a patched ffmpeg and then mpv with this PR, I figured I'd share how to resolve the conflict.

zhangn1985 commented:
from #14511 
to use this hwdec, command line is:
```mpv --hwdec=v4l2request --hwdec-codecs=all --gpu-hwdec-interop=drmprime-overlay --drm-draw-plane=overlay --drm-drmprime-video-plane=primary <hevc file>```

I will test this command line.

zhangn1985 commented:
`./mpv --gpu-hwdec-interop=v4l2request-overlay --drm-draw-plane=overlay --drm-drmprime-video-plane=primary /mnt/1.mp4 `

works good.

zhangn1985 commented:
@philipl I find Gstreamer already supports V4l2 request API directly without FFmpeg, would MPV merge direct v4l2 request Hwdec without ffmpeg?

philipl commented:
> @philipl I find Gstreamer already supports V4l2 request API directly without FFmpeg, would MPV merge direct v4l2 request Hwdec without ffmpeg?

gstreamer supporting a feature is the equivalent of ffmpeg supporting a feature, with respect to media players that use them. Your question would be the same as asking if GNOME Videos could directly use v4l2request instead of using it through gstreamer.

mpv will support it when ffmpeg supports it.

zhangn1985 commented:
> > @philipl I find Gstreamer already supports V4l2 request API directly without FFmpeg, would MPV merge direct v4l2 request Hwdec without ffmpeg?
> 
> gstreamer supporting a feature is the equivalent of ffmpeg supporting a feature, with respect to media players that use them. Your question would be the same as asking if GNOME Videos could directly use v4l2request instead of using it through gstreamer.
> 
> mpv will support it when ffmpeg supports it.

I understand, directly use V4l2 request API is not a player need to do.

Headcrabed commented:
@zhangn1985 What really needed is to get @Kwiboo 's [patch series](https://patchwork.ffmpeg.org/project/ffmpeg/cover/20240806090607.43240-1-jonas@kwiboo.se/) merged by upstream ffmpeg.

robertojguerra commented:
Hello all :)

I would like to help testing mpv with the v4l2request additional code.

I was able to compile mpv from https://github.com/philipl/mpv/tree/v4l2request , but I didn't get hw acceleration... it started using gpu-next

Then, in the Armbian forum, they suggested that I use mpv-0.3.9 with this pull request.

I was able to download the correct branch with:
git clone https://github.com/mpv-player/mpv
cd mpv
git checkout -b release/0.39
But at that point, **I don't know how to use git to download and use this PR 14690.**
Any help?
(Note: I would be testing it with armbian trixie xfce, and later labwc, with linux 6.15.4 in an orange pi zero 3)

Headcrabed commented:
Newest upstream progress here: https://code.ffmpeg.org/FFmpeg/FFmpeg/pulls/20847

robertojguerra commented:
In Debian Trixie, I use ffmpeg 7.1.x... do you know if the ffmpeg-v4l2request patch is appliable to ffmpeg 7.x.x?

Or more directly: is the code in https://code.ffmpeg.org/Kwiboo/FFmpeg/src/branch/v4l2request-v3 usable in Debian Trixie?

Kwiboo commented:
> I use ffmpeg 7.1.x... do you know if the ffmpeg-v4l2request patch is appliable to ffmpeg 7.x.x?

I do carry a copy of the main FFmpeg PR for `n7.1.3` and `n8.0.1` at https://code.ffmpeg.org/Kwiboo/FFmpeg/branches, all branches was just recently updated to include the `av_free()` fix that was added to the main PR.

chewi commented:
This was with ffmpeg 8.0.1, so Kwiboo's [v4l2-request-n8.0.1](https://code.ffmpeg.org/Kwiboo/FFmpeg/src/branch/v4l2-request-n8.0.1) branch, or you can take [LibreELEC's patch](https://github.com/LibreELEC/LibreELEC.tv/tree/master/packages/multimedia/ffmpeg/patches/v4l2-request) for this.

I thought you might also need lrusak's DRM PRIME patch as well, but I've just tested without, and it made no difference.

For mpv, I applied the changes from this PR.

I am a maintainer for ffmpeg in Gentoo, and up till now, we have only been conditionally applying jc-kynesim's patches for Raspberry Pi support. I had hoped that would benefit other boards too, but I now understand the difference between stateful and stateless VPUs. Unfortunately, jc-kynesim's and Kwiboo's patches collide, so LibreELEC only applies one or the other depending on the target. I didn't want to complicate Gentoo with that choice, so I looked into whether they could somehow be combined. They actually only collide in one area, the stateless HEVC decoder. In theory, jc-kynesim's should work beyond the Pi, so I went with his implementation and combined the rest. I haven't actually tested on a Pi and the HEVC VPU for the Quartz64 doesn't have mainline kernel support anyway, but I think it'll work. I'll hopefully get this into Gentoo soon.

robertojguerra commented:
Thank you for all the information.

I was just making up my mind about testing ffmeg-v4l2request + mpv in my orange pi zero 3 (allwinner h618) with Armbian trixie, which ships with ffmpeg 7 ... does anybody think ffmpeg 8 will work in Debian Trixie?

philipl commented:
In general, it works fine with vo=gpu-next. If it doesn't, it might be a problem with the GL or Vulkan driver for that hardware, but as long as it supports dmabuf imports, there's no reason it won't work.

philipl commented:
The empty list means that no modifiers are available.

chewi commented:
@Kwiboo My previously reported success was on a Quartz64 with H.264. There is no mainline HEVC decoder driver for that board, so I wasn't able to test that. I had hoped to test HEVC on the Pi 4 though, but I feel like I'm trying to get the stars to align.

With jc-kynesim's FFmpeg patches, I eventually managed to get Kodi to work. Even though his patches are supposed to support v4l2request for HEVC, it doesn't even appear under `mpv --hwdec=help`. Only v4l2m2m does, but that's no use for HEVC.

With [your FFmpeg patches](https://github.com/Kwiboo/FFmpeg/commits/v4l2-request-n8.1), with or without the additional v4l2-drmprime and vf-deinterlace-v4l2m2m patches, Kodi (GBM) doesn't show any video at all. mpv falls back to software decoding with this:

```
$ mpv -v --no-config --vo=gpu --gpu-context=drm --hwdec=v4l2request --opengl-es=yes --profile=fast --gpu-hwdec-interop=v4l2request-overlay

[vo/gpu/opengl] Initializing GPU context 'drm'
[vo/gpu/drm] VT_GETMODE failed: Inappropriate ioctl for device
[vo/gpu/drm] Failed to set up VT switcher. Terminal switching will be unavailable.
[vo/gpu/drm] Picked DRM card 0, primary node /dev/dri/card1 as the default.
[vo/gpu/drm] Driver: vc4 0.0.0 (0)
[vo/gpu/drm] Connector 35 currently connected to encoder 34
[vo/gpu/drm] Selected Encoder 34 with CRTC 102
[vo/gpu/drm] Selected mode: 2560x1440 (2560x1440@59.95Hz)
[vo/gpu/drm] DRM Atomic support found
[vo/gpu/drm] Using primary plane 91 as draw plane
[vo/gpu/drm] Using overlay plane 127 as drmprime plane
[vo/gpu] GBM_FORMAT_ARGB8888 supported by draw plane.
[vo/gpu] Supported modifier: 0x700000000000001
[vo/gpu] Supported modifier: 0x0
[vo/gpu] Creating GBM device
[vo/gpu] Initializing GBM surface (2560 x 1440)
[vo/gpu/opengl] Initializing EGL
[vo/gpu/opengl] EGL_VERSION=1.5
[vo/gpu/opengl] EGL_VENDOR=Mesa Project
[vo/gpu/opengl] EGL_CLIENT_APIS=OpenGL OpenGL_ES 
[vo/gpu/opengl] Trying to create GLES 2.x + context.
[vo/gpu/opengl] Attempting to find EGLConfig matching GBM_FORMAT_ARGB8888
[vo/gpu/opengl] Found matching EGLConfig for GBM_FORMAT_ARGB8888
[vo/gpu/opengl] Initializing EGL surface
[vo/gpu] GL_VERSION='OpenGL ES 3.1 Mesa 26.0.4'
[vo/gpu] Detected GLES 3.1.
[vo/gpu] GL_VENDOR='Broadcom'
[vo/gpu] GL_RENDERER='V3D 4.2.14.0'
[vo/gpu] GL_SHADING_LANGUAGE_VERSION='OpenGL ES GLSL ES 3.10'
[vo/gpu] Loaded extension GL_EXT_texture_norm16.
[vo/gpu] Loaded extension GL_EXT_color_buffer_half_float.
[vo/gpu] Loaded extension GL_EXT_disjoint_timer_query.
[vo/gpu] Loaded extension GL_EXT_buffer_storage.
[vo/gpu] Loaded extension GL_KHR_debug.
[vo/gpu/opengl] Preparing framebuffer
[vo/gpu/drm] crtc is not VRR capable
[vo/gpu/drm] Monitor pixel aspect: 1
[vo/gpu/opengl] Could not find path to render node.
[vo/gpu] Testing FBO format rgba16f
[vo/gpu] Using FBO format rgba16f.
[vo/gpu] No advanced processing required. Enabling dumb mode.
[vo/gpu] Loading hwdec driver 'v4l2request-overlay'
[vo/gpu/v4l2request-overlay] Using primary plane 91 as draw plane
[vo/gpu/v4l2request-overlay] Using overlay plane 127 as drmprime plane
[vo/gpu] Assuming 59.950550 FPS for display sync.
[vd] Container reported FPS: 25.000000
[vd] Codec list:
[vd]     hevc - HEVC (High Efficiency Video Coding)
[vd]     hevc_v4l2m2m (hevc) - V4L2 mem2mem HEVC decoder wrapper
[vd] Opening decoder hevc
[vd] Looking at hwdec hevc-v4l2request...
[vd] Trying hardware decoding via hevc-v4l2request.
[vd] Requesting 4 threads for decoding.
[vd] Selected decoder: hevc - HEVC (High Efficiency Video Coding)
[vf] User filter list:
[vf]   (empty)
[sub/lavc] Using subtitle decoder dvdsub
[cplayer] Starting playback...
[vd] Pixel formats supported by decoder: vulkan drm_prime yuv420p
[vd] Codec profile: Main (0x1)
[vd] Requesting pixfmt 'drm_prime' from decoder.
[ffmpeg/video] hevc: Failed setup for format drm_prime: hwaccel initialisation returned error.
[vd] Pixel formats supported by decoder: vulkan yuv420p yuv420p
[vd] Codec profile: Main (0x1)
[vd] Requesting pixfmt '-' from decoder.
[ffmpeg/video] hevc: Error parsing NAL unit #0.
[vd] Pixel formats supported by decoder: vulkan drm_prime yuv420p
[vd] Codec profile: Main (0x1)
[vd] Requesting pixfmt 'drm_prime' from decoder.
[ffmpeg/video] hevc: Failed setup for format drm_prime: hwaccel initialisation returned error.
[vd] Pixel formats supported by decoder: vulkan yuv420p yuv420p
[vd] Codec profile: Main (0x1)
[vd] Requesting pixfmt '-' from decoder.
[ffmpeg/video] hevc: Error parsing NAL unit #0.
[vd] Attempting next decoding method after failure of hevc-v4l2request.
[vd] Using software decoding.
[vd] Detected 4 logical cores.
[vd] Requesting 5 threads for decoding.
[vo/gpu] DR path suspected slow/uncached, disabling.
[vd] DR failed - disabling.
[vd] Using software decoding.
```

`--vo=gpu-next` always segfaults for me, even on the Quartz64.

```
[vo/gpu-next/opengl] Preparing framebuffer
[vo/gpu-next/drm] crtc is not VRR capable
[vo/gpu-next/drm] Monitor pixel aspect: 1
[vo/gpu-next/opengl] Could not find path to render node.
Segmentation fault
```

I have tried the latest published RPi 6.12, 6.18, and my own build of 6.18 using raspberrypi/linux@48abadbd4df1bc7e27ddb4798915d9ba8b8ab22f. I can see that the kernel bits changed in just the last few days, but it looks like this commit should work.

Where am I going wrong? I'm trying not to sound like a needy user here as I don't even really need this for myself. I'm just trying to get things working for Gentoo users.

chewi commented:
Figured that out. I was running the RPi kernel, but I'd built ffmpeg against vanilla headers. D'oh. Adding those two new pixel formats fixed Kodi. I get a blank screen with mpv though and this error repeatedly:

```
[vo/gpu] Failed to commit atomic request: Error number 22 occurred
```

chewi commented:
Aha, just had to add `--drm-draw-plane=overlay --drm-drmprime-video-plane=primary` to swap the planes. 9% CPU while playing a 720p HEVC, nice! Please let's get this merged.

Still no luck with gpu-next, but that's probably something else entirely.

kasper93 commented:
> Please let's get this merged.

This will be merged immediately after the upstream changes are finalized https://code.ffmpeg.org/FFmpeg/FFmpeg/pulls/20847

Even if public API would not change, you still currently need to build custom ffmpeg, so might as well build custom mpv linked to that.

diederikdehaas commented:
> When I build `libplacebo_7.351.0-352` I got a different failure, which may be useful (after all):
> 
> ```
> ----------------------------------- stderr -----------------------------------
> [0.001][w] API debugging requested but no debug meta layers present... ignoring
> [7.333][w] Requested swap mode unsupported by this device, falling back to VK_PRESENT_MODE_FIFO_KHR
> test.vulkan.c: ../src/vulkan/gpu_tex.c:419: vk_tex_create: Assertion `drm_list.drmFormatModifierCount > 0' failed.
> ==============================================================================
> 
> 
> Summary of Failures:
> 
> 12/16 libplacebo:opengl_surfaceless.c FAIL               4.13s   exit status 1
> 13/16 libplacebo:vulkan.c             FAIL               7.41s   killed by signal 6 SIGABRT
> 
> Ok:                14  
> Fail:              2   
> dh_auto_test: error: cd obj-aarch64-linux-gnu && DEB_PYTHON_INSTALL_LAYOUT=deb LC_ALL=C.UTF-8 MESON_TESTTHREADS=8 meson test --verbose -t 100 returned exit code 1
> make[1]: *** [debian/rules:12: override_dh_auto_test] Error 25
> make[1]: Leaving directory '/media/nvme/dev/debian/salsa/multimedia-team/libplacebo'
> make: *** [debian/rules:6: binary] Error 2
> dpkg-buildpackage: error: debian/rules binary subprocess failed with exit status 2
> ```
> 
> OTOH, building `libplacebo_7.351.0-354` did NOT fail the `vulkan.c` test, only the `opengl_surfaceless.c` one, which failed in every build I did. While it seems to have been fixed later on, the error message in -352 seems too 'good' to the reported issue, to be irrelevant.

Did another attempt to see if things have improved/fixed, but they are not :-/  
With ``libplacebo-dev`` version 7.360.1-1 I build a snapshot of ffmpeg + PR 20847 and with that I build a snapshot of mpv + patches from this PR, but I'm still getting the same failures.

I did learn a new 'trick' though, which resulted in https://salsa.debian.org/diederik/libplacebo/-/commits/bisect-libplacebo-352 and see the Salsa CI pipeline now succeeds.
But when I try to build that on my Rock 5B (RK3588) the build still fails due to the above quoted error.

So I tried disabling (ie 'bisecting') the patches/commits from libplacebo 351 till 352 to see which commit triggered the ``vulkan.c`` assertion and that turned out to be this commit:  
337eefe45350 ("vulkan/formats: add support for VK_KHR_format_feature_flags_2")

The build still failed due to the ``opengl_surfaceless.c`` problem with the previous commit, thus  
00a1009a7843 ("glsl/spirv_shaderc: use shaderc enum values for Vulkan version")

Building 7.360.1-1 on my Rock 5B fails because ``opengl_surfaceless.c`` fails, but it does not/no longer fail on ``vulkan.c``.
If I want to do some further 'bisecting' operation, then the ``opengl_surfaceless.c`` failure would need to be resolved so that I can at least build ``libplacebo`` (myself on the Rock 5B).

But I hope finding commit 337eefe45350 as the commit which first triggered the ``vulkan.c`` assertion may give a clue as to the failures I'm seeing.

diederikdehaas commented:
> > When I build `libplacebo_7.351.0-352` I got a different failure, which may be useful (after all):
> > ```
> > ----------------------------------- stderr -----------------------------------
> > [0.001][w] API debugging requested but no debug meta layers present... ignoring
> > [7.333][w] Requested swap mode unsupported by this device, falling back to VK_PRESENT_MODE_FIFO_KHR
> > test.vulkan.c: ../src/vulkan/gpu_tex.c:419: vk_tex_create: Assertion `drm_list.drmFormatModifierCount > 0' failed.
> > ==============================================================================
> > ...
> 
> So I tried disabling (ie 'bisecting') the patches/commits from libplacebo 351 till 352 to see which commit triggered the `vulkan.c` assertion and that turned out to be this commit:  
> 337eefe45350 ("vulkan/formats: add support for VK_KHR_format_feature_flags_2")

I just noticed the follow part of 337eefe45350's commit message:  
"As an aside, we are no longer at risk of truncating the modifier list"

So that seems like an incorrect assumption?
